feat: implement toString instances for Com&  Expr to enable opt tool to pretty print

salinhkuhn opened the pull request:
This PR implements toString functions for the Com and Expr types, enabling the opt tool to print the bb0.mlir file to the command line in a more readable format.
It also enhances the opt tool with a --verbose flag, which uses the Repr instances to print detailed output. If the flag is not provided, the output defaults to the toString representations.
Currently names are not preserved and the return statement is printed as a "return" not yet in the dialect generic syntax.
E.g the following input: 
```
  ^bb0(%arg0: i32):
    %0 = "llvm.mlir.constant"() {value = 8 : i32} : () -> i32
    %1 = "llvm.mlir.constant"() {value = 31 : i32} : () -> i32
    %2 = "llvm.or"(%0, %1) <{isDisjoint}> : (i32, i32) -> i32
    %3 = "llvm.sdiv"(%0, %2) <{isExact}> : (i32, i32) -> i32
    "llvm.return"(%3) : (i32) -> ()
```
will be printed to the command line as follows by invoking lake exe opt:
```
^entry(%0 : i32):
   %1 = "llvm.mlir.constant" { value = 8 : i32 } : () → (i32)
   %2 = "llvm.mlir.constant" { value = 31 : i32 } : () → (i32)
   %3 = "llvm.or disjoint" (%1, %2) : (i32, i32) → (i32)
   %4 = "llvm.sdiv exact" (%1, %3) : (i32, i32) → (i32)
  return %4 : (i32) → ()
```

tobiasgrosser committed:
chore: update to lean nightly 2025-05-13 (#1214)

automatic update of mathlib + lean via GitHub action.

Co-authored-by: github-merge-queue <118344674+github-merge-queue@users.noreply.github.com>

Commit:
breaks at the moment

Commit:
pretty printing of operands when using opt

Commit:
removed serialize again

Commit:
improved comments

Commit:
rmv serialized leftovers

tobiasgrosser commented on SSA/Projects/InstCombine/Base.lean:
```diff
@@ -386,6 +404,7 @@ namespace LLVM.Ty
 def width : LLVM.Ty → Nat
   | bitvec w => w
 
+
```
Unrelated.

tobiasgrosser replied:
Unrelated.

tobiasgrosser commented:
The CI fails with:

> info: Except.ok ⟨EffectKind.pure, ⟨i32, "llvm.and" (%3, %1) : (i32, i32) → (i32)⟩⟩
> info: /home/runner/work/lean-mlir/lean-mlir/SSA/Projects/InstCombine/Test.lean:179:15: Except.ok ⟨EffectKind.pure, ⟨i32, "llvm.add" (%4, %3) : (i32, i32) → (i32)⟩⟩
> error: /home/runner/work/lean-mlir/lean-mlir/SSA/Projects/InstCombine/Test.lean:179:0: ❌️ Docstring on `#guard_msgs` does not match generated message:

Also, this PR warns that the base branch is not up-to-date with the latest changes.

tobiasgrosser commented on SSA/Core/MLIRSyntax/GenericParser.lean:
```diff
@@ -350,9 +350,8 @@ info: MLIR.AST.MLIRType.undefined "shape.value"
-info: MLIR.AST.MLIRType.int (MLIR.AST.Signedness.Signless) (ConcreteOrMVar.concrete 32)
--/
+
+/-- info: MLIR.AST.MLIRType.int (MLIR.AST.Signedness.Signless) i32 -/
```
Is this change needed?

salinhkuhn replied:
yes, else it complains. It's because now we print i32 and not  (ConcreteOrMVar.concrete 32).

alexkeizer replied:
Should this not print either `i32` or `[mlir_type| i32]` (rather than MLIR.AST.MLIRType.int (MLIR.AST.Signedness.Signless) ... full-out), since the fact that it is a singless int is already implied by `i32`?

tobiasgrosser commented:
Can we have somehow a complete guard_msg test case that shows how these changes work. Similarly, it would be nice if the commit message would show the before/after of the `opt` output for our `bb` test case.

salinhkuhn commented:
> The CI fails with:
> 
> > info: Except.ok ⟨EffectKind.pure, ⟨i32, "llvm.and" (%3, %1) : (i32, i32) → (i32)⟩⟩
> > info: /home/runner/work/lean-mlir/lean-mlir/SSA/Projects/InstCombine/Test.lean:179:15: Except.ok ⟨EffectKind.pure, ⟨i32, "llvm.add" (%4, %3) : (i32, i32) → (i32)⟩⟩
> > error: /home/runner/work/lean-mlir/lean-mlir/SSA/Projects/InstCombine/Test.lean:179:0: ❌️ Docstring on `#guard_msgs` does not match generated message:
> 
> Also, this PR warns that the base branch is not up-to-date with the latest changes.

I just checked that and they fail because they expect the printing to be different. They require the output to be more verbose e.g : 

```Except.ok ⟨EffectKind.pure, ⟨i32, InstCombine.MOp.const (ConcreteOrMVar.concrete 32) 8 : () → (i32)⟩⟩```


 but with the new printing we get 

```Except.ok ⟨EffectKind.pure, ⟨i32, "llvm.mlir.constant" { value = 8 : i32 } : () → (i32)⟩```

salinhkuhn commented:
> Can we have somehow a complete guard_msg test case that shows how these changes work. Similarly, it would be nice if the commit message would show the before/after of the `opt` output for our `bb` test case.

I added before and after to the commit message now. I also pushed again and added a testcase to the beginning of the Test.lean.

Commit:
added test case to begin of test.lean

alexkeizer commented:
Firstly, `Repr` has the implicit contract that it's output is expected to be parse-able Lean code (resulting in the same object). It is thus not a good idea to use Repr for our specialized pretty-printing purposes. However, `ToString` has no such contract, and is totally fair game. Would it be possible to use only `ToString` for pretty printing, leaving the `Repr` instance as is? That way, there is not strictly a need for a new typeclass (yet): we already have the distinction between Repr and ToString.

That said, for LLVM, specifically, I'm curious how attributes are handled. Could you include an example program that has those, for example, how does a program with an `llvm.add %2, %0 overflow<nuw> : i32` instruction get pretty-printed? It's fine if they are ignored for now (after all, we are primarily interested in RiscV, which shouldn't have attributes), but, if that is the case, make sure you document this very explicitly (both in the PR description and in the code).

alexkeizer commented on SSA/Core/Util/ConcreteOrMVar.lean:
```diff
@@ -12,12 +12,28 @@ import Mathlib.Data.Vector.Basic
+  toString x :=
+    match x with
+    | .concrete a => s!"i{a}"
+    | .mvar i => s!"mvar({i})"
```
Do we need all these three instances? Could we leave the Repr instance as the derived one, and only use `ToString` for pretty-printing?

alexkeizer replied:
Note also that there is `ConcreteOrMVar` is not only used for signless-integer types. They could be used as the width of an unsigned int, or even as the value of a (non-negative) constant.

For example, IIUC, the type `u32` would be pretty-printed as:
`MLIR.AST.MLIRType.int (MLIR.AST.Signedness.Unsigned) i32`
which should be `u32` or `[mlir_type| u32]` instead!

I believe you should do the pretty printing at the `MLIRType` level instead, since that's where you can actually inspect the signedness and determine if you should prefix it with an `i` or `u` or `si` depending on said signedness.

salinhkuhn replied:
I got stuck there. If we could look at that in the call later that would be helpful. But I will look into it again now

salinhkuhn replied:
> Do we need all these three instances? Could we leave the Repr instance as the derived one, and only use `ToString` for pretty-printing?

I removed the explicit Repr instance and toFormat instance. Therefore only the toString instance is left.

alexkeizer commented on SSA/Projects/InstCombine/Base.lean:
```diff
@@ -99,7 +99,13 @@ def reprWithoutFlags (op : MOp.BinaryOp) (prec : Nat) : Format :=
+This affects every use of `Repr` for `MOp.BinaryOp`, which may not be desirable.
+TODO: Consider whether this change is acceptable globally.
+An alternative would be to introduce a separate `Serialize` typeclass that is same as `Repr` but uses the "llvm.{op}" format,
+while keeping `Repr` more verbose "InstCombine.MOp.BinaryOp.{op}". -/
```
As mentioned on Zulip: for `Repr` there is the expectation that the result is valid Lean code, so this is indeed not acceptable. However, `ToString` has no such expectation, so if you define a `ToString` instance like this and use that for pretty printing, that is totally fine!

alexkeizer commented on SSA/Projects/InstCombine/Base.lean:
```diff
@@ -115,6 +121,15 @@ inductive MOp (φ : Nat) : Type
+     | .select  _w => repr "select"
+     | .icmp  _pred _w => repr "icmp"
+     | .const w val => f!"\"llvm.mlir.constant\" \{ value = {val} : {w} }"
+
```
Here, too, could we make this the `ToString` instance rather than `Repr`?

alexkeizer reviewed (commented):
I think we should move away from `Repr` for pretty-printing, as we're breaking some of the implicit `Repr` contract with these instances! In future we could add a new DialectSerialize typeclass, but for now ToString should suffice, so let's try that first.

alexkeizer committed:
fix: add no_index to bind_if_{then,else}_poison lemmas (#1216)

These lemmas had a mention of `PoisonOr` in their discrimination tree
key, since ite has an implicit argument for it's return type. By adding
no_index, we fix this, meaning we no longer have to put `(...)` around
this lemmas when using it in the simp_llvm simpset.

salinhkuhn committed:
feat: hybrid RiscV plus LLVM dialect [1/3] (#1212)

This PR introduces the core definitions for the Hybrid Dialect, which
merges functionality from both the RV64 and InstCombine dialects. It
includes:
-Definitions of types and operands.
-Implementations of typeclass instances for the dialect.
-functions to assist with dialect parsing, specifically for
simultaneously mapping values and types.
The full parsing infrastructure will be included in a follow-up PR.

---------

Co-authored-by: Sarah Kuhn <kuhnsa@student.ethz.ch>
Co-authored-by: Alex Keizer <alex@keizer.dev>
Co-authored-by: Tobias Grosser <tobias@grosser.es>

salinhkuhn committed:
fix: remove transform error type dependency (#1217)

This PR removes the type dependency from TransformError. Instead of
taking an input argument of type d.Ty, TransformError now takes no
arguments. This change was necessary to enable staged parsing, as
previously each dialect had its own ExceptM monad. By removing the
dependency from TransformError, it's no longer required for each dialect
to throw exceptions using its own monad.
To support this change, the PR adds ToString instances to existing
projects such as the FHE and the PaperExample/Simple dialects.

---------

Co-authored-by: Sarah Kuhn <kuhnsa@student.ethz.ch>
Co-authored-by: Alex Keizer <alex@keizer.dev>

tobiasgrosser committed:
chore: disable unused variable error (#1221)

tobiasgrosser committed:
chore: shorten proof (#1222)

This is a pure cleanup with no function change intended, that was pulled
out of https://github.com/opencompl/lean-mlir/pull/1211/files, as it was
unrelated to this PR.

alexkeizer committed:
refactor: replace alive_auto with bv_auto in AliveStatements (#1218)

alive_auto is a tactic that runs the whole stack of `simp_alive_{ops,
undef, etc.}` simplification tactics before finally calling out to
`bv_auto` (the actual bitvec level solver tactic). Since we already run
through the simplification stack of tactics in AliveStatements, the
latter is thus much more appropriate.

Commit:
wip: toString instances for com and expr

tobiasgrosser committed:
chore: reduce ci threads to 48 to make CI more robust (#1223)

salinhkuhn committed:
To int to nat removal semantics: RISCV Merge [1/3] (#1211)

This is a rather big PR again but it contains the semantics already
merged and additionally has rewrites of the semantics into purely bit
vector domain statements without calls to toInt and toNat. This file
should replace the current smenatics file when complete.

This PR is the first third of the whole undertaking.
For some theorems, we previously already committed their pure variant
but left out the toNat version. In this case, we now add the toNat
version under the original theorem name, move the pure version to a _bv
statement, and prove equivalence of both.

---------

Co-authored-by: Sarah Kuhn <kuhnsa@student.ethz.ch>
Co-authored-by: Tobias Grosser <tobias@grosser.es>
Co-authored-by: Luisa Cicolini <48860705+luisacicolini@users.noreply.github.com>

Commit:
Merge remote-tracking branch 'origin/main' into minimal-Op-printing-Repr

Commit:
fix: Repr

tobiasgrosser commented on SSA/Core/Framework.lean:
```diff
@@ -267,9 +267,48 @@ def Lets.repr (prec : Nat) : Lets d eff Γ t → Format
 instance : Repr (Expr d Γ eff t) := ⟨flip Expr.repr⟩
 instance : Repr (Com d Γ eff t) := ⟨flip Com.repr⟩
 instance : Repr (Lets d Γ eff t) := ⟨flip Lets.repr⟩
-
```
unrelated

tobiasgrosser commented on SSA/Core/MLIRSyntax/GenericParser.lean:
```diff
@@ -350,9 +350,8 @@ info: MLIR.AST.MLIRType.undefined "shape.value"
-info: MLIR.AST.MLIRType.int (MLIR.AST.Signedness.Signless) (ConcreteOrMVar.concrete 32)
--/
+
+/-- info: MLIR.AST.MLIRType.int (MLIR.AST.Signedness.Signless) (ConcreteOrMVar.concrete 32) -/
```
I think you can avoid this diff by keeping the original formatting.

tobiasgrosser commented on SSA/Core/Util/ConcreteOrMVar.lean:
```diff
@@ -5,19 +5,34 @@ import Mathlib.Data.Fin.Basic
 import Mathlib.Data.Vector.Basic
 -- import Mathlib.Tactic
 
-
```
unrelated.

salinhkuhn commented:
So I implemented part of the toString functions for Expr and Com. When I run 
lake exe opt on bb0.mlir

I get the following output: 
```
{ 
^entry(%0 : i32):
%1 = "llvm.mlir.constant" { value = 8 : i32 } : () → (i32)
%2 = "llvm.mlir.constant" { value = 31 : i32 } : () → (i32)
%3 = "llvm.ashr" (%0, %2) : (i32, i32) → (i32)
%4 = "llvm.and" (%3, %1) : (i32, i32) → (i32)
%5 = "llvm.add" (%4, %3) : (i32, i32) → (i32)
return %5 : (i32) → ()
 }
 ```
 if we pass the --verbose flag, it should print using the Repr and output :
 ```
Flag `--verbose` was set.
⟨EffectKind.pure, ⟨i32, {
    ^entry(%0 : i32):
      %1 = InstCombine.MOp.const (ConcreteOrMVar.concrete 32) 8 : () → (i32)
      %2 = InstCombine.MOp.const (ConcreteOrMVar.concrete 32) 31 : () → (i32)
      %3 = InstCombine.MOp.binary
        (ConcreteOrMVar.concrete 32)
        (InstCombine.MOp.BinaryOp.ashr) (%0, %2) : (i32, i32) → (i32)
      %4 = InstCombine.MOp.binary
        (ConcreteOrMVar.concrete 32)
        (InstCombine.MOp.BinaryOp.and) (%3, %1) : (i32, i32) → (i32)
      %5 = InstCombine.MOp.binary
        (ConcreteOrMVar.concrete 32)
        (InstCombine.MOp.BinaryOp.add) (%4, %3) : (i32, i32) → (i32)
      return %5 : (i32) → ()
  }
```
Attributes especially flags are not working yet. But it would be nice to get a quick opinion if that's a path to follow further.

tobiasgrosser commented on SSA/Projects/InstCombine/Test.lean:
```diff
@@ -108,6 +119,7 @@ info: Except.ok ⟨EffectKind.pure, ⟨i32, InstCombine.MOp.const (ConcreteOrMVa
 -/
 #guard_msgs in #eval mkExpr    (Γn 2) op1    ["0", "arg0"]
 
+#eval mkExpr    (Γn 3) op2    ["1", "0", "arg0"]
```
These changes are not needed, no?

salinhkuhn replied:
No, I shall remove them. It was to figure out why the CI failed on the previous push.

tobiasgrosser commented:
> So I implemented part of the toString functions for Expr and Com. When I run lake exe opt on bb0.mlir
> 
> I get the following output:
> 
> ```
> { 
> ^entry(%0 : i32):
> %1 = "llvm.mlir.constant" { value = 8 : i32 } : () → (i32)
> %2 = "llvm.mlir.constant" { value = 31 : i32 } : () → (i32)
> %3 = "llvm.ashr" (%0, %2) : (i32, i32) → (i32)
> %4 = "llvm.and" (%3, %1) : (i32, i32) → (i32)
> %5 = "llvm.add" (%4, %3) : (i32, i32) → (i32)
> return %5 : (i32) → ()
>  }
> ```
> 
> if we pass the --verbose flag, it should print using the Repr and output :
> ```

Nice. I left some comments. Otherwise, this LGTM, but let's see what @alexkeizer says.

You could also indent the `%***` lines by a bit, making the overall code look nicer.

Commit:
Merge remote-tracking branch 'origin/main' into minimal-Op-printing-Repr

alexkeizer commented on SSA/Core/Framework.lean:
```diff
@@ -267,9 +267,48 @@ def Lets.repr (prec : Nat) : Lets d eff Γ t → Format
+/-- Format a list of formal arguments as `(%0 : t₀, %1 : t₁, ... %n : tₙ)` -/
+partial def formatFormalArgListTupleStr [ToString Ty] (ts : List Ty) : String :=
+  let args := (List.range ts.length).zip ts |>.map
+    (fun (i, t) => s!"%{i} : {toString t}") -- assumes toString function for types. -- tried using it with naming but then difficult to track.
```
I'm not sure what this comment is saying, could you explain a bit more (doesn't have to be as a comment, feel free to just reply to this comment)?

Also, do keep the max line width in mind: that applies to comments, too!

salinhkuhn replied:
I experimented with different ways of naming variables—specifically, I tried  naming free variables with arg0 to argn, and then assigning new local variables with purely numeric names. However, this approach caused issues: when one of the free variables needed to be printed, it ended up being assigned a numeric name as well, which led to confusion and unexpected behavior. Overall, I explored various naming strategies, but they didn't work out as intended

alexkeizer commented on SSA/Core/Framework.lean:
```diff
@@ -267,9 +267,48 @@ def Lets.repr (prec : Nat) : Lets d eff Γ t → Format
+    (fun (i, t) => s!"%{i} : {toString t}") -- assumes toString function for types. -- tried using it with naming but then difficult to track.
+  "(" ++ String.intercalate ", " args ++ ")"
+
+  partial def Expr.toString1 [ToString d.Op] : Expr d Γ eff t → String
```
Why is this called `toString1`, and not just `toString`?
```suggestion
  partial def Expr.toString [ToString d.Op] : Expr d Γ eff t → String
```

salinhkuhn replied:
Whenever I rename the function to just toString, the calls to {toString op} and {toString outTy} result in errors. Lean tries to resolve toString as the one defined for Expr, so when I pass something of type d.Op or d.Ty, it complains because it expects an Expr. However, when I change the function name to something else, like toString1, the issue disappears

salinhkuhn replied:
resolved now.

alexkeizer commented on SSA/Core/Framework.lean:
```diff
@@ -267,9 +267,48 @@ def Lets.repr (prec : Nat) : Lets d eff Γ t → Format
+
+  partial def Expr.toString1 [ToString d.Op] : Expr d Γ eff t → String
+    | Expr.mk (op : d.Op) _ _ args regArgs =>
+        let outTy : d.Ty  := DialectSignature.outTy op
```
```suggestion
        let outTy : d.Ty := DialectSignature.outTy op
```
mind your whitespace (check this thoroughly, please!)

alexkeizer commented on SSA/Core/Framework.lean:
```diff
@@ -267,9 +267,48 @@ def Lets.repr (prec : Nat) : Lets d eff Γ t → Format
+
+      s!"{toString op}{formatArgTuple args} : {formatTypeTuple argTys} → ({toString outTy})"
+
+partial def comToStringAux : Com d Γ eff t → String
```
Why not:
```suggestion
partial def Com.toStringAux : Com d Γ eff t → String
```
That would be a bit more consistent. You could even go a bit more descriptive and name it `Com.toStringBody` to emphasize it's formatting just the body of the basic block.

alexkeizer commented on SSA/Core/Framework.lean:
```diff
@@ -267,9 +267,48 @@ def Lets.repr (prec : Nat) : Lets d eff Γ t → Format
+
+
+
+
```
```suggestion
```
why so many newlines?

alexkeizer commented on SSA/Core/Util/ConcreteOrMVar.lean:
```diff
@@ -5,19 +5,34 @@ import Mathlib.Data.Fin.Basic
+  reprPrec x _prec :=
+    match x with
+    | .concrete a => f!"i{a}"
+    | .mvar i => f!"mvar({i})" -/
```
Could you remove this instance, rather than leaving it commented out?
```suggestion
```

salinhkuhn replied:
done.

alexkeizer commented on SSA/Projects/InstCombine/Base.lean:
```diff
@@ -100,6 +100,12 @@ def reprWithoutFlags (op : MOp.BinaryOp) (prec : Nat) : Format :=
+This affects every use of `Repr` for `MOp.BinaryOp`, which may not be desirable.
+TODO: Consider whether this change is acceptable globally.
+An alternative would be to introduce a separate `Serialize` typeclass that is same as `Repr` but uses the "llvm.{op}" format,
+while keeping `Repr` more verbose "InstCombine.MOp.BinaryOp.{op}". -/
```
Is this comment still up-to-date and accurate? If not, we should remove it

salinhkuhn replied:
removed it.

alexkeizer commented on SSA/Projects/InstCombine/Base.lean:
```diff
@@ -115,6 +121,45 @@ inductive MOp (φ : Nat) : Type
+instance : ToString (MOp 0) where
+   toString  op :=
+     match op with
+     | .unary _w op => s!"\"{repr op}\""
```
Should this not be `toString` rather than `repr`?
```suggestion
     | .unary _w op => s!"\"{toString op}\""
```

salinhkuhn replied:
Done. I use the Repr instance for the unary operations when defining their toString. Since the Repr instance is derived and doesn't include the InstCombine prefix, the output is already clean and readable, making it suitable to reuse directly for toString.

alexkeizer reviewed (commented):
It already looks good, although it seems you still have some old comments lying around that need cleaning up.

Also, could you add docstrings to your definitions to explain what they do? We haven't really been great at documenting everything previously, but for new PRs we should strive to have a docstring for each new definition.

Finally, try to format your PR title as a proper sentence, so put spaces between words rather than dots and dashes, and use imperative style (read [mathlib's PR style guide](https://leanprover-community.github.io/contribute/commit.html) for more detail; although we aren't always as strict as the mathlib guideline, it'd be good to build a habit of following what they prescribe in that document)

Commit:
fix: adressed review comments, improved doc

Commit:
fix: removed leftover comment

Commit:
rm unrelated changes

Commit:
rm unrelated changes

Commit:
formating of changes

Commit:
fix whitespace

Commit:
removed wellformedness printing comment

salinhkuhn commented:
To visualize what this PR does: 
The input file with the following content, will be printed to the command line by running lake exe opt "file-path". 

```
  ^bb0(%arg0: i32):
    %0 = "llvm.mlir.constant"() {value = 8 : i32} : () -> i32
    %1 = "llvm.mlir.constant"() {value = 31 : i32} : () -> i32
    %2 = "llvm.ashr"(%arg0, %1) : (i32, i32) -> i32
    %3 = "llvm.and"(%2, %0) : (i32, i32) -> i32
    %4 = "llvm.add"(%3, %2) : (i32, i32) -> i32
    "llvm.return"(%4) : (i32) -> ()
```
The output will be the following:
```
^entry(%0 : i32):
  %1 = "llvm.mlir.constant" { value = 8 : i32 } : () → (i32)
  %2 = "llvm.mlir.constant" { value = 31 : i32 } : () → (i32)
  %3 = "llvm.ashr" (%0, %2) : (i32, i32) → (i32)
  %4 = "llvm.and" (%3, %1) : (i32, i32) → (i32)
  %5 = "llvm.add" (%4, %3) : (i32, i32) → (i32)
 return %5 : (i32) → ()
```
Programs containing flags e.g 
```
  ^bb0(%arg0: i32):
    %0 = "llvm.mlir.constant"() {value = 8 : i32} : () -> i32
    %1 = "llvm.mlir.constant"() {value = 31 : i32} : () -> i32
    %2 = "llvm.or"(%0, %1) <{isDisjoint}> : (i32, i32) -> i32
    %3 = "llvm.sdiv"(%0, %2) <{isExact}> : (i32, i32) -> i32
    "llvm.return"(%3) : (i32) -> ()
```
will be displayed as follows :
```
^entry(%0 : i32):
   %1 = "llvm.mlir.constant" { value = 8 : i32 } : () → (i32)
   %2 = "llvm.mlir.constant" { value = 31 : i32 } : () → (i32)
   %3 = "llvm.or disjoint" (%1, %2) : (i32, i32) → (i32)
   %4 = "llvm.sdiv exact" (%1, %3) : (i32, i32) → (i32)
  return %4 : (i32) → ()
```
Additionally, one can pass the --verbose flag to get more detailed output, including the effectKind, which will be printed as follows (for the first example):
```
Flag `--verbose` was set.
⟨EffectKind.pure, ⟨i32, {
    ^entry(%0 : i32):
      %1 = InstCombine.MOp.const (ConcreteOrMVar.concrete 32) 8 : () → (i32)
      %2 = InstCombine.MOp.const (ConcreteOrMVar.concrete 32) 31 : () → (i32)
      %3 = InstCombine.MOp.binary
        (ConcreteOrMVar.concrete 32)
        (InstCombine.MOp.BinaryOp.ashr) (%0, %2) : (i32, i32) → (i32)
      %4 = InstCombine.MOp.binary
        (ConcreteOrMVar.concrete 32)
        (InstCombine.MOp.BinaryOp.and) (%3, %1) : (i32, i32) → (i32)
      %5 = InstCombine.MOp.binary
        (ConcreteOrMVar.concrete 32)
        (InstCombine.MOp.BinaryOp.add) (%4, %3) : (i32, i32) → (i32)
      return %5 : (i32) → ()
  }⟩⟩
```

alexkeizer commented:
> Finally, try to format your PR title as a proper sentence, so put spaces between words rather than dots and dashes, and use imperative style (read [mathlib's PR style guide](https://leanprover-community.github.io/contribute/commit.html) for more detail; although we aren't always as strict as the mathlib guideline, it'd be good to build a habit of following what they prescribe in that document)

@salinhkuhn did you have a look at the PR style guide I linked in my previous review?

alexkeizer commented:
> To visualize what this PR does: The input file with the following content, will be printed to the command line by running lake exe opt "file-path".
> 
> ```
>   ^bb0(%arg0: i32):
>     %0 = "llvm.mlir.constant"() {value = 8 : i32} : () -> i32
>     %1 = "llvm.mlir.constant"() {value = 31 : i32} : () -> i32
>     %2 = "llvm.ashr"(%arg0, %1) : (i32, i32) -> i32
>     %3 = "llvm.and"(%2, %0) : (i32, i32) -> i32
>     %4 = "llvm.add"(%3, %2) : (i32, i32) -> i32
>     "llvm.return"(%4) : (i32) -> ()
> ```
> 
> The output will be the following:
> 
> ```
> ^entry(%0 : i32):
>   %1 = "llvm.mlir.constant" { value = 8 : i32 } : () → (i32)
>   %2 = "llvm.mlir.constant" { value = 31 : i32 } : () → (i32)
>   %3 = "llvm.ashr" (%0, %2) : (i32, i32) → (i32)
>   %4 = "llvm.and" (%3, %1) : (i32, i32) → (i32)
>   %5 = "llvm.add" (%4, %3) : (i32, i32) → (i32)
>  return %5 : (i32) → ()
> ```
> 
> Programs containing flags e.g
> 
> ```
>   ^bb0(%arg0: i32):
>     %0 = "llvm.mlir.constant"() {value = 8 : i32} : () -> i32
>     %1 = "llvm.mlir.constant"() {value = 31 : i32} : () -> i32
>     %2 = "llvm.or"(%0, %1) <{isDisjoint}> : (i32, i32) -> i32
>     %3 = "llvm.sdiv"(%0, %2) <{isExact}> : (i32, i32) -> i32
>     "llvm.return"(%3) : (i32) -> ()
> ```
> 
> will be displayed as follows :
> 
> ```
> ^entry(%0 : i32):
>    %1 = "llvm.mlir.constant" { value = 8 : i32 } : () → (i32)
>    %2 = "llvm.mlir.constant" { value = 31 : i32 } : () → (i32)
>    %3 = "llvm.or disjoint" (%1, %2) : (i32, i32) → (i32)
>    %4 = "llvm.sdiv exact" (%1, %3) : (i32, i32) → (i32)
>   return %4 : (i32) → ()
> ```
> 
> Additionally, one can pass the --verbose flag to get more detailed output, including the effectKind, which will be printed as follows (for the first example):
> 
> ```
> Flag `--verbose` was set.
> ⟨EffectKind.pure, ⟨i32, {
>     ^entry(%0 : i32):
>       %1 = InstCombine.MOp.const (ConcreteOrMVar.concrete 32) 8 : () → (i32)
>       %2 = InstCombine.MOp.const (ConcreteOrMVar.concrete 32) 31 : () → (i32)
>       %3 = InstCombine.MOp.binary
>         (ConcreteOrMVar.concrete 32)
>         (InstCombine.MOp.BinaryOp.ashr) (%0, %2) : (i32, i32) → (i32)
>       %4 = InstCombine.MOp.binary
>         (ConcreteOrMVar.concrete 32)
>         (InstCombine.MOp.BinaryOp.and) (%3, %1) : (i32, i32) → (i32)
>       %5 = InstCombine.MOp.binary
>         (ConcreteOrMVar.concrete 32)
>         (InstCombine.MOp.BinaryOp.add) (%4, %3) : (i32, i32) → (i32)
>       return %5 : (i32) → ()
>   }⟩⟩
> ```

Thanks for writing this summary, it helps a lot!

The first thing that jumps out at me is the fact that 
```
"llvm.return"(%3) : (i32) -> ()
```
gets printed as 
```
return %5 : (i32) → ()
```

Since we seem to be printing using the generic syntax for operations, we should keep that convention and use the generic syntax for printing returns also. That is, the output should be the same as the input on those examples you gave (modulo the attributes).

By the way, when asking for a second round of review, it's also good to check your PR description again, and make sure it's still accurate and up-to-date. In particular, I believe we've moved away from the global Repr instance modifications, right? So please update the PR description to reflect that.
The PR description end up in the git history for this PR, so it's important that the description matches the actual changes that get merged at the end accurately!

alexkeizer commented on SSA/Core/Framework.lean:
```diff
@@ -270,6 +270,47 @@ instance : Repr (Lets d Γ eff t) := ⟨flip Lets.repr⟩
 
+/- # ToString instances for Com and Expr  -/
+section ToString
+variable {d} [DialectSignature d] [Repr d.Op] [Repr d.Ty][ToString d.Ty][ToString d.Op]
```
```suggestion
variable {d} [DialectSignature d] [Repr d.Op] [Repr d.Ty] [ToString d.Ty] [ToString d.Op]
```

Keep a space in between binders!

alexkeizer commented on SSA/Core/Framework.lean:
```diff
@@ -270,6 +270,47 @@ instance : Repr (Lets d Γ eff t) := ⟨flip Lets.repr⟩
+-/
+partial def Expr.toString [ToString d.Op] : Expr d Γ eff t → String
+  | Expr.mk (op : d.Op) _ _ args _regArgs =>
+    let outTy : d.Ty  := DialectSignature.outTy op
```
```suggestion
    let outTy : d.Ty := DialectSignature.outTy op
```

alexkeizer commented on SSA/Core/Framework.lean:
```diff
@@ -270,6 +270,47 @@ instance : Repr (Lets d Γ eff t) := ⟨flip Lets.repr⟩
+    Com.ToStringBody body
+
+/- This function implements a toString instance for the type `Com`.  -/
+partial def Com.toString (_com : Com d Γ eff t) : String :=
```
```suggestion
/- `Com.toString` implements a toString instance for the type `Com`.  -/
partial def Com.toString (_com : Com d Γ eff t) : String :=
```

refer to the function being documented by name, instead of "this function" (check this in other defs also, please!)

alexkeizer commented on SSA/Core/Framework.lean:
```diff
@@ -270,6 +270,47 @@ instance : Repr (Lets d Γ eff t) := ⟨flip Lets.repr⟩
+partial def Com.toString (_com : Com d Γ eff t) : String :=
+   "{ \n"
+  ++ "^entry" ++  ((formatFormalArgListTupleStr Γ)) ++ ":" ++ "\n"
+  ++ (Com.ToStringBody _com) ++
```
```suggestion
  ++ (Com.ToStringBody com) ++
```
if you use the variable, it shouldn't have the underscore

alexkeizer commented on SSA/Projects/InstCombine/Base.lean:
```diff
@@ -115,6 +115,47 @@ inductive MOp (φ : Nat) : Type
   | const (w : Width φ) (val : ℤ) : MOp φ
 deriving Repr, DecidableEq, Inhabited, Lean.ToExpr
 
+def toStringWithoutFlags (op : MOp.BinaryOp) : String :=
```
This is called `withoutFlags`, but actually it does have the flags!

I think this way of printing the attributes is good for now, so that at least they are present, but indeed in future we would remove those attributes/flags from this function once we have proper attribute support (so that this function really becomes "withoutFlags". In the present, though, the name is confusing. Either we change the name now (and change it back to withoutFlags in the future), or we keep it as-is, but then you have to add a comment (preferably as a docstring) explaining that *at the moment* it does include the flags, but this will be changed in the future.

salinhkuhn replied:
I implemented the change and renamed the function to toStringWithFlags, so whenever we add attribute support in the future we can change it.

alexkeizer commented on SSA/Projects/InstCombine/LLVM/Opt.lean:
```diff
@@ -1,19 +1,38 @@
 import Cli
 
+
+def verbose_flag (fileName : String ) : IO UInt32 := do
+      let icom? ← parseComFromFile fileName
+      match icom? with
+      | none => return 1
+      | some (Sigma.mk _Γ icom) => do
+      IO.println "Flag `--verbose` was set."
+      IO.println s!"{repr icom}" -- we print everything, meaning effect, return type and com
+      return 0
```
```suggestion
def verboseFlag (fileName : String ) : IO UInt32 := do
      let icom? ← parseComFromFile fileName
      match icom? with
      | none => return 1
      | some (Sigma.mk _Γ icom) => do
      IO.println "Flag `--verbose` was set."
      IO.println s!"{repr icom}" -- we print everything, meaning effect, return type and com
      return 0
```

Check the [naming convention](https://leanprover-community.github.io/contribute/naming.html): regular definitions like this are `lowerCamelCase`.

Also, please add a docstring, in particular, to explain (a) that this function will print stuff (as that's not obvious from the name), and (b) what the meaning of the returned u32 is (I presume it's meant as an exit code?; if so, no need to document each value, but do document that it *is* an exit code and that 0 means success)

salinhkuhn replied:
done

alexkeizer commented on SSA/Projects/InstCombine/LLVM/Opt.lean:
```diff
@@ -1,19 +1,38 @@
     opt VIA runMainCmd;
     "opt: apply verified rewrites"
+    FLAGS:
+      verbose; "Declares a flag `--verbose`. This is the description of the flag."
```
I think you forgot to replace the example string with an actual description here?

alexkeizer reviewed (approved):
Some last comments, but once those (and the two comments preceding this top-level one!) are addressed feel free to merge. Do remember to write docstrings for new definitions!

salinhkuhn committed:
Update SSA/Core/Framework.lean

Co-authored-by: Alex Keizer <alex@keizer.dev>

salinhkuhn committed:
Update SSA/Core/Framework.lean

Co-authored-by: Alex Keizer <alex@keizer.dev>

salinhkuhn committed:
Update SSA/Core/Framework.lean

Co-authored-by: Alex Keizer <alex@keizer.dev>

salinhkuhn committed:
Update SSA/Core/Framework.lean

Co-authored-by: Alex Keizer <alex@keizer.dev>

salinhkuhn commented:
> > To visualize what this PR does: The input file with the following content, will be printed to the command line by running lake exe opt "file-path".
> > ```
> >   ^bb0(%arg0: i32):
> >     %0 = "llvm.mlir.constant"() {value = 8 : i32} : () -> i32
> >     %1 = "llvm.mlir.constant"() {value = 31 : i32} : () -> i32
> >     %2 = "llvm.ashr"(%arg0, %1) : (i32, i32) -> i32
> >     %3 = "llvm.and"(%2, %0) : (i32, i32) -> i32
> >     %4 = "llvm.add"(%3, %2) : (i32, i32) -> i32
> >     "llvm.return"(%4) : (i32) -> ()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The output will be the following:
> > ```
> > ^entry(%0 : i32):
> >   %1 = "llvm.mlir.constant" { value = 8 : i32 } : () → (i32)
> >   %2 = "llvm.mlir.constant" { value = 31 : i32 } : () → (i32)
> >   %3 = "llvm.ashr" (%0, %2) : (i32, i32) → (i32)
> >   %4 = "llvm.and" (%3, %1) : (i32, i32) → (i32)
> >   %5 = "llvm.add" (%4, %3) : (i32, i32) → (i32)
> >  return %5 : (i32) → ()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Programs containing flags e.g
> > ```
> >   ^bb0(%arg0: i32):
> >     %0 = "llvm.mlir.constant"() {value = 8 : i32} : () -> i32
> >     %1 = "llvm.mlir.constant"() {value = 31 : i32} : () -> i32
> >     %2 = "llvm.or"(%0, %1) <{isDisjoint}> : (i32, i32) -> i32
> >     %3 = "llvm.sdiv"(%0, %2) <{isExact}> : (i32, i32) -> i32
> >     "llvm.return"(%3) : (i32) -> ()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > will be displayed as follows :
> > ```
> > ^entry(%0 : i32):
> >    %1 = "llvm.mlir.constant" { value = 8 : i32 } : () → (i32)
> >    %2 = "llvm.mlir.constant" { value = 31 : i32 } : () → (i32)
> >    %3 = "llvm.or disjoint" (%1, %2) : (i32, i32) → (i32)
> >    %4 = "llvm.sdiv exact" (%1, %3) : (i32, i32) → (i32)
> >   return %4 : (i32) → ()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Additionally, one can pass the --verbose flag to get more detailed output, including the effectKind, which will be printed as follows (for the first example):
> > ```
> > Flag `--verbose` was set.
> > ⟨EffectKind.pure, ⟨i32, {
> >     ^entry(%0 : i32):
> >       %1 = InstCombine.MOp.const (ConcreteOrMVar.concrete 32) 8 : () → (i32)
> >       %2 = InstCombine.MOp.const (ConcreteOrMVar.concrete 32) 31 : () → (i32)
> >       %3 = InstCombine.MOp.binary
> >         (ConcreteOrMVar.concrete 32)
> >         (InstCombine.MOp.BinaryOp.ashr) (%0, %2) : (i32, i32) → (i32)
> >       %4 = InstCombine.MOp.binary
> >         (ConcreteOrMVar.concrete 32)
> >         (InstCombine.MOp.BinaryOp.and) (%3, %1) : (i32, i32) → (i32)
> >       %5 = InstCombine.MOp.binary
> >         (ConcreteOrMVar.concrete 32)
> >         (InstCombine.MOp.BinaryOp.add) (%4, %3) : (i32, i32) → (i32)
> >       return %5 : (i32) → ()
> >   }⟩⟩
> > ```
> 
> Thanks for writing this summary, it helps a lot!
> 
> The first thing that jumps out at me is the fact that
> 
> ```
> "llvm.return"(%3) : (i32) -> ()
> ```
> 
> gets printed as
> 
> ```
> return %5 : (i32) → ()
> ```
> 
> Since we seem to be printing using the generic syntax for operations, we should keep that convention and use the generic syntax for printing returns also. That is, the output should be the same as the input on those examples you gave (modulo the attributes).
> 
> By the way, when asking for a second round of review, it's also good to check your PR description again, and make sure it's still accurate and up-to-date. In particular, I believe we've moved away from the global Repr instance modifications, right? So please update the PR description to reflect that. The PR description end up in the git history for this PR, so it's important that the description matches the actual changes that get merged at the end accurately!

I tried to experiment with preserving the input and printing exactly the input to the command line again but did not see a way to properly do that. Especially preserving the name of the free variables seemed hard. I can try to improve that in the future.
Your point regarding the return statement is a good one, but to implement that I would need a toString instance for every dialect. But since the dialect is not an inductive type but a structure, this change requires a bigger change in the framework, I suppose e.g the dialect structure could have a field of type String that asks for a name. 

update: we decided to implement to proper return statement printing and name preservation in a separate PR in the future.

Commit:
fix comment

Commit:
fix: implemeted feedback

tobiasgrosser commented:
> > > To visualize what this PR does: The input file with the following content, will be printed to the command line by running lake exe opt "file-path".
> > > ```
> > >   ^bb0(%arg0: i32):
> > >     %0 = "llvm.mlir.constant"() {value = 8 : i32} : () -> i32
> > >     %1 = "llvm.mlir.constant"() {value = 31 : i32} : () -> i32
> > >     %2 = "llvm.ashr"(%arg0, %1) : (i32, i32) -> i32
> > >     %3 = "llvm.and"(%2, %0) : (i32, i32) -> i32
> > >     %4 = "llvm.add"(%3, %2) : (i32, i32) -> i32
> > >     "llvm.return"(%4) : (i32) -> ()
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > The output will be the following:
> > > ```
> > > ^entry(%0 : i32):
> > >   %1 = "llvm.mlir.constant" { value = 8 : i32 } : () → (i32)
> > >   %2 = "llvm.mlir.constant" { value = 31 : i32 } : () → (i32)
> > >   %3 = "llvm.ashr" (%0, %2) : (i32, i32) → (i32)
> > >   %4 = "llvm.and" (%3, %1) : (i32, i32) → (i32)
> > >   %5 = "llvm.add" (%4, %3) : (i32, i32) → (i32)
> > >  return %5 : (i32) → ()
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Programs containing flags e.g
> > > ```
> > >   ^bb0(%arg0: i32):
> > >     %0 = "llvm.mlir.constant"() {value = 8 : i32} : () -> i32
> > >     %1 = "llvm.mlir.constant"() {value = 31 : i32} : () -> i32
> > >     %2 = "llvm.or"(%0, %1) <{isDisjoint}> : (i32, i32) -> i32
> > >     %3 = "llvm.sdiv"(%0, %2) <{isExact}> : (i32, i32) -> i32
> > >     "llvm.return"(%3) : (i32) -> ()
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > will be displayed as follows :
> > > ```
> > > ^entry(%0 : i32):
> > >    %1 = "llvm.mlir.constant" { value = 8 : i32 } : () → (i32)
> > >    %2 = "llvm.mlir.constant" { value = 31 : i32 } : () → (i32)
> > >    %3 = "llvm.or disjoint" (%1, %2) : (i32, i32) → (i32)
> > >    %4 = "llvm.sdiv exact" (%1, %3) : (i32, i32) → (i32)
> > >   return %4 : (i32) → ()
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Additionally, one can pass the --verbose flag to get more detailed output, including the effectKind, which will be printed as follows (for the first example):
> > > ```
> > > Flag `--verbose` was set.
> > > ⟨EffectKind.pure, ⟨i32, {
> > >     ^entry(%0 : i32):
> > >       %1 = InstCombine.MOp.const (ConcreteOrMVar.concrete 32) 8 : () → (i32)
> > >       %2 = InstCombine.MOp.const (ConcreteOrMVar.concrete 32) 31 : () → (i32)
> > >       %3 = InstCombine.MOp.binary
> > >         (ConcreteOrMVar.concrete 32)
> > >         (InstCombine.MOp.BinaryOp.ashr) (%0, %2) : (i32, i32) → (i32)
> > >       %4 = InstCombine.MOp.binary
> > >         (ConcreteOrMVar.concrete 32)
> > >         (InstCombine.MOp.BinaryOp.and) (%3, %1) : (i32, i32) → (i32)
> > >       %5 = InstCombine.MOp.binary
> > >         (ConcreteOrMVar.concrete 32)
> > >         (InstCombine.MOp.BinaryOp.add) (%4, %3) : (i32, i32) → (i32)
> > >       return %5 : (i32) → ()
> > >   }⟩⟩
> > > ```
> > 
> > 
> > Thanks for writing this summary, it helps a lot!
> > The first thing that jumps out at me is the fact that
> > ```
> > "llvm.return"(%3) : (i32) -> ()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > gets printed as
> > ```
> > return %5 : (i32) → ()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Since we seem to be printing using the generic syntax for operations, we should keep that convention and use the generic syntax for printing returns also. That is, the output should be the same as the input on those examples you gave (modulo the attributes).
> > By the way, when asking for a second round of review, it's also good to check your PR description again, and make sure it's still accurate and up-to-date. In particular, I believe we've moved away from the global Repr instance modifications, right? So please update the PR description to reflect that. The PR description end up in the git history for this PR, so it's important that the description matches the actual changes that get merged at the end accurately!
> 
> I tried to experiment with preserving the input and printing exactly the input to the command line again but did not see a way to properly do that. Especially preserving the name of the free variables seemed hard. I can try to improve that in the future. Your point regarding the return statement is a good one, but to implement that I would need a toString instance for every dialect. But since the dialect is not an inductive type but a structure, this change requires a bigger change in the framework, I suppose e.g the dialect structure could have a field of type String that asks for a name.

I think that's fine for now and we may want to keep it the way it is in this PR. There is generally no expectation that names are preserved. We can refine the return statement as it becomes a problem.

Commit:
fix: rm dialect toStrig attempt

alexkeizer commented:
> Especially preserving the name of the free variables seemed hard. I can try to improve that in the future.

Sorry, I should have been more precise there, as @tobiasgrosser mentioned, the names of the variables aren't important

tobiasgrosser commented:
> > Finally, try to format your PR title as a proper sentence, so put spaces between words rather than dots and dashes, and use imperative style (read [mathlib's PR style guide](https://leanprover-community.github.io/contribute/commit.html) for more detail; although we aren't always as strict as the mathlib guideline, it'd be good to build a habit of following what they prescribe in that document)
> 
> @salinhkuhn did you have a look at the PR style guide I linked in my previous review?

Sarah, did you look at this. In particular the pr title seems not yet compliant.

salinhkuhn commented:
> > > Finally, try to format your PR title as a proper sentence, so put spaces between words rather than dots and dashes, and use imperative style (read [mathlib's PR style guide](https://leanprover-community.github.io/contribute/commit.html) for more detail; although we aren't always as strict as the mathlib guideline, it'd be good to build a habit of following what they prescribe in that document)
> > 
> > 
> > @salinhkuhn did you have a look at the PR style guide I linked in my previous review?
> 
> Sarah, did you look at this. In particular the pr title seems not yet compliant.

I'm just now reading it and taking notes. I'm doing a PR learning session now for the next hour such that future PR's (there will be quite a few thinking of my instructions lowering) are better quality and easier to review.

Not sure if that's needed except for me but I could also write a few points about it into the knowledge base.
